From fbf4d64c5cb1e949a3d2c221b2a0cbf8603c5313 Mon Sep 17 00:00:00 2001
From: Luciano Pacheco <lucmult@chromium.org>
Date: Thu, 12 Sep 2024 05:00:40 +0000
Subject: [PATCH] SelectFileDialogExtension: Fix termination to always cleanup

In rare occasions the dialog and its delegate have the termination
initiated by the aura::Window, instead of content::WebContents.
See both stack traces in the bug comment #13.

When this happens the views::WidgetDelegate::WindowWillClose() isn't
called. The SelectFileDialogExtension and its delegate
SystemFilesAppDialogDelegate relied on `WindowWillClose()` to perform
their cleanup. Thus, the cleanup wasn't performed causing a crash.

This CL changes the cleanup to be based on:
WebDialogDelegate::RegisterOnDialogClosedCallback() which fixes the
issue, this callback is triggered for both termination paths.

Bug: 361042814
Change-Id: Icde5174fb119860bccefebefaab059382ac10af8
Cq-Do-Not-Cancel-Tryjobs: true
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5832772
Commit-Queue: Luciano Pacheco <lucmult@chromium.org>
Reviewed-by: Austin Tankiang <austinct@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1354353}
---

--- a/chrome/browser/ui/views/select_file_dialog_extension.cc
+++ b/chrome/browser/ui/views/select_file_dialog_extension.cc
@@ -200,7 +200,11 @@
                                std::u16string title)
       : ash::SystemWebDialogDelegate(url, title),
         id_(id),
-        parent_(std::move(parent)) {}
+        parent_(std::move(parent)) {
+    RegisterOnDialogClosedCallback(
+        base::BindOnce(&SystemFilesAppDialogDelegate::OnDialogClosing,
+                       base::Unretained(this)));
+  }
   ~SystemFilesAppDialogDelegate() override = default;
 
   void SetModal(bool modal) {
@@ -231,7 +235,7 @@
     ash::SystemWebDialogDelegate::OnDialogShown(webui);
   }
 
-  void OnDialogWillClose() override {
+  void OnDialogClosing(const std::string& ret_val) {
     if (parent_) {
       parent_->OnSystemDialogWillClose();
     }
@@ -504,6 +508,7 @@
   owner_ = owner;
   type_ = type;
 
+  // The delegate deletes itself in WebDialogDelegate::OnDialogClosed().
   auto* dialog_delegate = new SystemFilesAppDialogDelegate(
       weak_factory_.GetWeakPtr(), routing_id, file_manager_url, dialog_title);
   dialog_delegate->SetModal(owner.window != nullptr);
